File: frontend/src/types/lease-term-version.ts
/**
 * Lease Term Version Types
 *
 * Zod schemas and TypeScript types for versioned lease recovery terms.
 * Each version has an effective date; the calculation engine uses the
 * version effective during the reconciliation period.
 */
import { z } from 'zod'

/**
 * Full lease term version (from database)
 */
export const LeaseTermVersionSchema = z.object({
  id: z.string().uuid(),
  lease_id: z.string().uuid(),
  version_number: z.number().int().min(1),
  effective_date: z.string(),

  // Recovery profile fields
  base_year: z.number().int().min(1990).max(2100).nullable().optional(),
  base_year_amount: z.string().nullable().optional(),
  gross_up_base_year: z.boolean().default(false),
  pro_rata_share: z.string(),
  cap_type: z
    .enum(['none', 'non_cumulative', 'cumulative', 'cumulative_compounding'])
    .default('none'),
  cap_rate: z.string().nullable().optional(),
  admin_fee_percentage: z.string().default('0'),
  excluded_pools: z.array(z.string()).default([]),
  rsf_measurement_standard: z.string().nullable().optional(),
  rsf_measurement_date: z.string().nullable().optional(),

  // Amendment metadata
  amendment_reason: z.string().nullable().optional(),
  amendment_document_url: z.string().max(2048).nullable().optional(),

  // Audit
  created_by: z.string().uuid().nullable().optional(),
  created_at: z.string(),
})

export type LeaseTermVersion = z.infer<typeof LeaseTermVersionSchema>

/**
 * DTO for creating a new term version (amendment)
 */
export const LeaseTermVersionCreateSchema = z
  .object({
    effective_date: z.string().min(1, 'Effective date is required'),
    base_year: z.number().int().min(1990).max(2100).nullable().optional(),
    base_year_amount: z.string().nullable().optional(),
    gross_up_base_year: z.boolean().optional(),
    pro_rata_share: z.string().min(1, 'Pro-rata share is required'),
    cap_type: z
      .enum(['none', 'non_cumulative', 'cumulative', 'cumulative_compounding'])
      .default('none'),
    cap_rate: z.string().nullable().optional(),
    admin_fee_percentage: z.string().optional(),
    excluded_pools: z.array(z.string()).optional(),
    amendment_reason: z.string().nullable().optional(),
    amendment_document_url: z.string().max(2048).nullable().optional(),
  })
  .refine(
    (data) => {
      if (data.cap_type && data.cap_type !== 'none') {
        return data.cap_rate && data.cap_rate !== ''
      }
      return true
    },
    {
      message: 'Cap rate is required when cap type is not none',
      path: ['cap_rate'],
    }
  )

export type LeaseTermVersionCreate = z.infer<
  typeof LeaseTermVersionCreateSchema
>

/**
 * Lightweight summary for timeline display
 */
export const LeaseTermVersionSummarySchema = z.object({
  id: z.string().uuid(),
  version_number: z.number().int(),
  effective_date: z.string(),
  pro_rata_share: z.string(),
  cap_type: z.string(),
  amendment_reason: z.string().nullable().optional(),
  created_at: z.string(),
})

export type LeaseTermVersionSummary = z.infer<
  typeof LeaseTermVersionSummarySchema
>

File: backend/app/models/lease_term_version.py
"""Lease Term Version models for temporal versioning of recovery terms.

Each lease can have multiple term versions with effective dates.
The calculation engine looks up the version effective during the
reconciliation period, enabling accurate historical reconciliations
even after terms change.
"""

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class LeaseTermVersion(BaseModel):
    """Full lease term version from database."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    lease_id: UUID
    version_number: int = Field(ge=1)
    effective_date: date

    # Recovery profile fields
    base_year: int | None = Field(None, ge=1990, le=2100)
    base_year_amount: Decimal | None = Field(None, ge=Decimal("0"))
    gross_up_base_year: bool = False
    pro_rata_share: Decimal = Field(..., ge=Decimal("0"), le=Decimal("1"))
    cap_type: str = Field(default="none")
    cap_rate: Decimal | None = Field(None, ge=Decimal("0"), le=Decimal("1"))
    admin_fee_percentage: Decimal = Field(
        default=Decimal("0"), ge=Decimal("0"), le=Decimal("0.20")
    )
    management_fee_percentage: Decimal | None = Field(
        default=None, ge=Decimal("0"), le=Decimal("0.20")
    )
    excluded_pools: list[str] = Field(default_factory=list)
    rsf_measurement_standard: str | None = None
    rsf_measurement_date: date | None = None

    # Amendment metadata
    amendment_reason: str | None = None
    amendment_document_url: str | None = Field(None, max_length=2048)

    # Audit
    created_by: UUID | None = None
    created_at: datetime

    @model_validator(mode="after")
    def validate_cap_rate_required(self) -> "LeaseTermVersion":
        """cap_rate must be provided when cap_type is not 'none'."""
        if self.cap_type != "none" and self.cap_rate is None:
            raise ValueError("cap_rate is required when cap_type is not none")
        return self


class LeaseTermVersionCreate(BaseModel):
    """DTO for creating a new term version (amendment)."""

    effective_date: date
    base_year: int | None = Field(None, ge=1990, le=2100)
    base_year_amount: Decimal | None = Field(None, ge=Decimal("0"))
    gross_up_base_year: bool = False
    pro_rata_share: Decimal = Field(..., ge=Decimal("0"), le=Decimal("1"))
    cap_type: str = Field(default="none")
    cap_rate: Decimal | None = Field(None, ge=Decimal("0"), le=Decimal("1"))
    admin_fee_percentage: Decimal = Field(
        default=Decimal("0"), ge=Decimal("0"), le=Decimal("0.20")
    )
    management_fee_percentage: Decimal | None = Field(
        default=None, ge=Decimal("0"), le=Decimal("0.20")
    )
    excluded_pools: list[str] = Field(default_factory=list)
    rsf_measurement_standard: str | None = None
    rsf_measurement_date: date | None = None
    amendment_reason: str | None = None
    amendment_document_url: str | None = Field(None, max_length=2048)

    @model_validator(mode="after")
    def validate_cap_rate_required(self) -> "LeaseTermVersionCreate":
        """cap_rate must be provided when cap_type is not 'none'."""
        if self.cap_type != "none" and self.cap_rate is None:
            raise ValueError("cap_rate is required when cap_type is not none")
        return self


class LeaseTermVersionSummary(BaseModel):
    """Lightweight model for timeline display."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    version_number: int
    effective_date: date
    pro_rata_share: Decimal
    cap_type: str
    amendment_reason: str | None = None
    created_at: datetime
